src/persist/src/indexed/trace.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A persistent, compacting data structure of `(Key, Value, Time, Diff)`
+//! updates, indexed by key.
+//!
+//! This is directly a persistent analog of [differential_dataflow::trace::Trace].
+
+use std::sync::Arc;
+
+use differential_dataflow::trace::Description;
+use timely::progress::Antichain;
+use timely::progress::Timestamp;
+
+use crate::error::Error;
+use crate::indexed::cache::BlobCache;
+use crate::indexed::BlobTraceBatch;
+use crate::indexed::BlobTraceMeta;
+use crate::persister::Snapshot;
+use crate::storage::Blob;
+
+/// A persistent, compacting data structure containing `(Key, Value, Time,
+/// Diff)` entries indexed by `(key, value, time)`.
+///
+/// Invariants:
+/// - All entries are before some time frontier.
+/// - This acts as an append-only log. Data is added in advancing batches and
+///   logically immutable after that (modulo compactions, which preserve it, but
+///   the ability to read at old times is lost).
+/// - TODO: Explain since and logical compactions.
+/// - TODO: Space usage.
+pub struct BlobTrace {
+    since: Antichain<u64>,
+    // NB: The Descriptions here are sorted and contiguous.
+    batches: Vec<(Description<u64>, String)>,
+}
+
+impl Default for BlobTrace {
+    fn default() -> Self {
+        BlobTrace::new(BlobTraceMeta {
+            since: Antichain::from_elem(Timestamp::minimum()),
+            batches: Vec::new(),
+        })
+    }
+}
+
+impl BlobTrace {
+    /// Returns a BlobTrace re-instantiated with the previously serialized
+    /// state.
+    pub fn new(meta: BlobTraceMeta) -> Self {
+        BlobTrace {
+            since: meta.since,
+            batches: meta.batches,
+        }
+    }
+
+    /// Serializes the state of this BlobTrace for later re-instantiation.
+    pub fn meta(&self) -> BlobTraceMeta {
+        BlobTraceMeta {
+            since: self.since.clone(),
+            batches: self.batches.clone(),
+        }
+    }
+
+    /// An upper bound on the times of contained updates.
+    pub fn frontier(&self) -> Antichain<u64> {
+        match self.batches.last() {
+            Some((desc, _)) => desc.upper().clone(),
+            None => Antichain::from_elem(Timestamp::minimum()),
+        }
+    }
+
+    /// A lower bound on the time at which updates may have been logically
+    /// compacted together.
+    pub fn since(&self) -> Antichain<u64> {
```

Comment:
should this be equal to the lowest `since` across all of the batch descriptions (and sinces are nondecreasing across batches)? (just want to double check my understanding of the invariant here)

Reply:
thought about this some more - `join`s in this context correspond to `max`s and I think the max since of all the batches makes sense. Like concretely: 

if we had a batch from [0, 5) since 3
and a second batch from [5, 10) since 7 -- then afaict the since of the overall trace is 7 (technically we can also answer questions about [3, 5) but not 6